Threading fallback - PR 1 - Link events with their related event

MidhunSureshR opened the pull request:
- [x] Create proper event-entries with event data fetched from hs
- [x] Decryption of event-entries created with data from hs
- [x] Display-name/avatar data with event-entries created with data from hs
- [ ] Log operations
## Overview:
- Implement context endpoint needed to fetch an event from hs
- Link every entry with its context-entry (taking care of things like updates)

## Terminologies / changes added to EventEntry:  

For a given `EventEntry` e:
- e.**contextEventId**: is the event-id of another event to which this event is a reply of.
- e.**contextEntry**: is the actual event-entry to which this event is a reply of.
- e.**contextForEntries**: = `{event| event.contextEntry = e}` i.e this is an array of event-entries such that for each event-entry in this array, *e* is its context.  
Used for propagating updates from a given entry to other entries that depend on it to render reply-previews; eg: if an entry is redacted, all other entries which use this entry as context for rendering reply preview must be notified (so that redaction can be rendered in reply previews).

## Terminologies / changes added to Timeline:  
- **contextEntriesNotInTimeline**: Is a Map of entries that were fetched from storage/hs because it was needed as a context for some other entry. They are fetched from storage/hs because they are currently not in the timeline( i.e not in remoteEntries ).

MidhunSureshR committed:
WIP

MidhunSureshR committed:
Formatting fix

MidhunSureshR committed:
await on loading related events

bwindels commented on src/matrix/room/BaseRoom.js:
```diff
@@ -501,7 +501,9 @@ export class BaseRoom extends EventEmitter {
                 logger: this._platform.logger,
-                powerLevelsObservable: await this.observePowerLevels()
+                powerLevelsObservable: await this.observePowerLevels(),
+                fetchEventFromStorage: eventId => this._readEventById(eventId),
```
why inject this? the homeserver already has access to storage ...

MidhunSureshR replied:
`_readEventById()` is already defined in BaseRoom (and is used in `BaseRoom.observeEvent()`?
We wouldn't want to duplicate the same logic in Timeline?

bwindels replied:
Ahh, yes, I see now. This takes care of decryption and everything :+1:

bwindels replied:
Hmm, thinking about this a bit more, I wonder if we should move _readEventById into `TimelineReader` as a public method? In _readEventById in here we would just create a new ephemeral TimelineReader to read a single event, and in the timeline we have a TimelineReader as well. TimelineReader already takes care of decryption as well.

bwindels commented on src/matrix/room/BaseRoom.js:
```diff
@@ -501,7 +501,9 @@ export class BaseRoom extends EventEmitter {
-                powerLevelsObservable: await this.observePowerLevels()
+                powerLevelsObservable: await this.observePowerLevels(),
+                fetchEventFromStorage: eventId => this._readEventById(eventId),
+                fetchEventFromHomeserver: eventId => this._getEventFromHomeserver(eventId)
```
I think we can also consider to pass hsApi into timeline for this... unless there's a good reason to inject? Unit tests perhaps?

MidhunSureshR replied:
Felt more cohesive to have this code in the same place that we have the code to fetch an event from storage. Would encapsulating these two methods into an `eventFetcher` be more agreeable?

bwindels replied:
So if we end up with readEventById in TimelineReader, I would put this in the Timeline. I don't thing we will ever really need to call /context from outside of the Timeline.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -211,8 +214,9 @@ export class Timeline {
-    replaceEntries(entries) {
+    async replaceEntries(entries) {
         this._addLocalRelationsToNewRemoteEntries(entries);
+        await this._loadRelatedEvents(entries);
```
I'm slightly worried about this blocking getting the initial event add/update out, Especially in the worst case scenario where we hit the homeserver to get the context event ... Perhaps we could not await this call and send an update (e.g. call replaceEntries again) once we have obtained the context event.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -234,10 +238,32 @@ export class Timeline {
-    addEntries(newEntries) {
+    async addEntries(newEntries) {
         this._addLocalRelationsToNewRemoteEntries(newEntries);
+        await this._loadRelatedEvents(newEntries);
```
Same here, slightly worried about blocking here.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -234,10 +238,32 @@ export class Timeline {
+    async _loadRelatedEvents(entries) {
+        const filteredEntries = entries.filter(e => !!e.relation);
+        for (const entry of filteredEntries) {
+            const id = entry.relatedEventId;
```
Two things:
 - the id can be undefined/null
 - I don't think we should fetch the event for every relation. This would include for example the reacted-to-event for reactions (which are present at this stage in the timeline, but filtered out in TilesCollection later on). I would add a separate property (e.g. `contextEventId`, adding a comment on top of the getter to say what it is for) that only returns the event id in case of a reply.

bwindels commented on src/matrix/room/timeline/entries/EventEntry.js:
```diff
@@ -41,6 +42,10 @@ export class EventEntry extends BaseEventEntry {
         }
     }
 
+    setRelatedEntry(entry) {
```
Wrt to the earlier comment, perhaps this can be called contextEntry?

bwindels reviewed (commented):
good start! first quick review, left some comments

MidhunSureshR committed:
Implement context endpoint

MidhunSureshR committed:
Parse display name and avatar of event

MidhunSureshR committed:
Use 'context' instead of 'related'

MidhunSureshR committed:
Add contextEventId

MidhunSureshR committed:
Change comment

MidhunSureshR committed:
Use NonPersistedEventEntry

MidhunSureshR committed:
Emit updated entries

MidhunSureshR commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -261,6 +261,8 @@ export class Timeline {
             if (contextEvent) {
                 entry.setContextEntry(contextEvent);
+                // emit this change
+                this._remoteEntries.update(entry);
```
I'm not sure if this is safe to do... is there a race between sortedArray operations done by `replaceEntries`/`addEntries` and this update :thinking: 
Maybe we need to do:
```js
this._remoteEntries.getAndUpdate(entry, (entry, existingEntry) => existingEntry);
```

MidhunSureshR replied:
I'm using this for now:
```js
this._findAndUpdateRelatedEntry(null, entry.id, () => true); 
```

MidhunSureshR committed:
Support decryption on entries fetched from hs

MidhunSureshR committed:
Support redaction changes in remoteEntries

MidhunSureshR committed:
Track fetched entries for redactions

MidhunSureshR committed:
Improve comment

MidhunSureshR committed:
findAndUpdate instead of update

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,8 +247,50 @@ export class Timeline {
+    }
+
+    async _loadRelatedEvents(entries) {
+        const filteredEntries = entries.filter(e => !!e.contextEventId);
```
```suggestion
        const entriesNeedingContext = entries.filter(e => !!e.contextEventId);
```

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,8 +247,50 @@ export class Timeline {
+                if (needToTrack) {
+                    // this entry was created from storage/hs, so it's not tracked by remoteEntries
+                    // we track them here so that we can update reply preview of dependents on redaction
+                    this._fetchedEventEntries.push(contextEvent);
```
If this only needs to happen if the entry comes from storage, can we move this below `await this._fetchEventFromStorage(id);` on line 275? Looks like we wouldn't need the `needToTrack` flag then? If you can replace the flag with control flow, that might be more readable.

OTOH, I wonder why we wouldn't need to track if the event came from the homeserver? Isn't it possible an event we don't have in storage gets redacted?

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,8 +247,50 @@ export class Timeline {
+            }
+            // fetch from hs
+            if (!contextEvent) {
+                contextEvent = await this._fetchEventFromHomeserver(id);
```
Shouldn't we also look into `_fetchedEventEntries` entries here first?

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,8 +247,50 @@ export class Timeline {
+                contextEvent.addDependent(entry);
+                entry.setContextEntry(contextEvent);
+                // emit this change
+                this._findAndUpdateRelatedEntry(null, entry.id, () => true);
```
```suggestion
                const txnId = entry.isPending? entry.id : null;
                const eventId = entry.isPending? null : entry.id;
                this._findAndUpdateRelatedEntry(txnId, eventId, () => true);
```

bwindels replied:
Can we rename _findAndUpdateRelatedEntry to _findAndUpdateEntryById (and the local variables called related... therein)? It doesn't really do anything specific for relations and the name is a bit confusing here.

bwindels replied:
Also, you could put the above in a `_updateEntry` method.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,8 +247,50 @@ export class Timeline {
+            // find in storage
+            if (!contextEvent) {
+                needToTrack = true;
+                contextEvent = await this._fetchEventFromStorage(id);
```
perhaps we could also use this in _getOrLoadEntry?

bwindels commented on src/matrix/room/BaseRoom.js:
```diff
@@ -559,6 +562,23 @@ export class BaseRoom extends EventEmitter {
+    async _getEventFromHomeserver(eventId) {
+        const response = await this._hsApi.context(this._roomId, eventId, 0).response();
+        const sender = response.event.sender;
+        const member = response.state.find(e => e.type === "m.room.member" && e.user_id === sender);
```
you can import the "m.room.member" string from "./members/RoomMember.js" if you wish.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,8 +247,50 @@ export class Timeline {
+
+    _updateFetchedEntries(entries) {
+        for (const entry of entries) {
+            const relatedEntry = this._fetchedEventEntries.find(e => e.id === entry.relatedEventId);
```
Should _fetchedEventEntries be a `Map` rather? Seems we would only need to look things up by id and never iterate otherwise?

bwindels reviewed (commented):
I'm thinking of the nomenclature introduced by this PR, so far:
 - context event: the event entry on a related entry that we need to display ourselves (on a reply)
 - dependants: the inverse relation of context events, e.g. target events that are mentioned as a context event on their relations (replies), needed to notify the dependent entries that their context event has changed.
 - fetched entries: context events that are not in the timeline, and tracked separatedly so they can trigger an update in their dependents when they get updated (e.g redacted)

Perhaps we could change it to:
 - stick with the term context (event)
 - rename dependants to `contextForEntries`
 - rename fetched entries to `contextEntriesNotInTimeline` or `contextEntriesOutOfScope`?

I'm also thinking:
 - we should somehow document (comments or elsewhere) this
 -  if we can possibly come up with something a bit more coherent

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,8 +247,50 @@ export class Timeline {
+    _updateFetchedEntries(entries) {
+        for (const entry of entries) {
+            const relatedEntry = this._fetchedEventEntries.find(e => e.id === entry.relatedEventId);
+            if (relatedEntry?.addLocalRelation(entry)) {
```
this is to make fetched event appear as redacted (and later edited) when we get a redaction? perhaps add a comment saying as much.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,8 +247,50 @@ export class Timeline {
+        for (const entry of entries) {
+            const relatedEntry = this._fetchedEventEntries.find(e => e.id === entry.relatedEventId);
+            if (relatedEntry?.addLocalRelation(entry)) {
+                relatedEntry.dependents.forEach(e => this._findAndUpdateRelatedEntry(null, e.id, () => true));
```
Could use the elsewhere proposed `_updateEntry` method here.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,8 +247,50 @@ export class Timeline {
     /** @package */
     addEntries(newEntries) {
         this._addLocalRelationsToNewRemoteEntries(newEntries);
+        this._updateFetchedEntries(newEntries);
```
addEntries is not just called for sync, but also for gap fills. So it is possible that an event we previously didn't know about has now been added to the storage and timeline by a gap fill. Should we transition the fetched entry to a remote event entry from the timeline in all of the dependants?

MidhunSureshR replied:
Yes we should. Otherwise the dependents of the entry created before gap fill will become dead/forgotten.

bwindels reviewed (commented):
some more thoughts/questions

MidhunSureshR committed:
Made code more readable

MidhunSureshR committed:
Use map and fetch from Map if available

MidhunSureshR committed:
Implement readById() in TimelineReader

MidhunSureshR committed:
Move methods into Timeline

MidhunSureshR committed:
Refactor out magic string

MidhunSureshR committed:
Use _updateEntry

MidhunSureshR committed:
Add explaining comment

MidhunSureshR committed:
Move event to remoteEntries if needed

MidhunSureshR committed:
Reuse code in getOrLoadEntry

MidhunSureshR committed:
Rename dependents --> contextForEntries

MidhunSureshR committed:
Rename fetchedEntries --> contextEntriesNotInTimeline

MidhunSureshR committed:
Check if contextEvent was found

MidhunSureshR committed:
Add tests

MidhunSureshR committed:
Fix bug

MidhunSureshR committed:
Remove entry from map

MidhunSureshR committed:
Propagate updates

MidhunSureshR committed:
Make tests pass

MidhunSureshR committed:
Rename methods

MidhunSureshR committed:
Remove dead code

MidhunSureshR committed:
No need to delete before update on map

MidhunSureshR committed:
Add test for move code

MidhunSureshR committed:
Update comment

MidhunSureshR committed:
Remove await from tests

MidhunSureshR committed:
Also copy over contextEntry from otherEntry

MidhunSureshR committed:
Update comment

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,7 +250,109 @@ export class Timeline {
+        }
+    }
+
+    _createEntryFromRelatedEntries(entry, relatedEntry) {
```
This seems unused?

MidhunSureshR replied:
It's used in [line 270, Timeline.js](https://github.com/vector-im/hydrogen-web/pull/611/files#diff-95a8d7280cef101142b1363df405903543be8531502e1b4aee38f208fb9a0ac0R270)

bwindels commented on src/matrix/room/timeline/entries/EventEntry.js:
```diff
@@ -39,6 +42,27 @@ export class EventEntry extends BaseEventEntry {
+        this._contextForEntries.push(entry);
+    }
+
+    setAsRedacted() {
```
If `_createEntryFromRelatedEntries` is unused, this is unused as well

bwindels replied:
I'm still wondering if adding a third way to mark an event as redacted is the best way to go here. Also, we only really need this method in `NonPersistedEventEntry` (as context entries not in the timeline but fetched from storage would get redacted by the relation writer and be passed to replaceEntries). So how about we add a `redact` method there (perhaps with a comment saying that redaction needs a method because non persisted entries don't go through relation writer) that runs the [redaction algorithm](https://github.com/vector-im/hydrogen-web/blob/master/src/matrix/room/timeline/persistence/RelationWriter.js#L130-L143) (which will need to be extracted from `RelationWriter`, perhaps to room/common.ts?) on a copy of the event and creates a new NonPersistedEventEntry with it.

Then we can get rid of this method and the flag.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,7 +250,109 @@ export class Timeline {
+        }
+    }
+
+    _getTrackedEntry(id) {
```
what does tracked mean here? is there perhaps a more appropriate name we can pick here?

bwindels replied:
is id always an eventId? if so, consider renaming it

MidhunSureshR replied:
> what does tracked mean here? is there perhaps a more appropriate name we can pick here?

tracked as in we track two kinds of entries:
- **remoteEntries** for entries in the timeline
- **contextEntriesNotInTimeline** for entries we had to manually fetch to render replies

I've added a jsdoc comment saying as much.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,7 +250,109 @@ export class Timeline {
+        this._findAndUpdateEntryById(txnId, eventId, () => true);
+    }
+
+    async _loadRelatedEvents(entries) {
```
I feel this name is both too vague and not using the terminology used elsewhere ("context" rather than "related"), which makes it hard to guess what this method does when seeing it used elsewhere in this class. What about `_loadContextEntriesWhereNeeded`?

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -43,6 +46,10 @@ export class Timeline {
         this._allEntries = null;
+        /** Stores event entries that we had to fetch from hs/storage for reply previews (because they were not in timeline) */ 
+        this._contextEntriesNotInTimeline = new Map();
+        this._decryptEntries = null;
```
perhaps add a comment that this is only used for decrypting non-persisted context entries loaded from the HS?

bwindels reviewed (commented):
seems like there is a merge conflict, and if I'm not missing anything some unused code. Perhaps best to go over it one last time to see everything is ready for final review?

MidhunSureshR committed:
Add jsdoc comment

MidhunSureshR committed:
Change method name

MidhunSureshR committed:
add more jsdoc comments

MidhunSureshR committed:
Add explaining comment

bwindels commented on src/matrix/room/timeline/entries/EventEntry.js:
```diff
@@ -24,6 +24,9 @@ export class EventEntry extends BaseEventEntry {
         this._decryptionError = null;
         this._decryptionResult = null;
+        this._contextEntry = null;
+        this._contextForEntries = null;
```
I'm wondering if we can avoid putting a new member variable in this class for a very specific use case (which tends to have a lot of instances, and I'd like to keep light). What about that instead of putting it here, we put the array of entries we are context for into `Timeline._contextEntriesNotInTimeline`, by changing the value from an entry to an object (of a new class, e.g. `ContextEntryLinks` or something) with the context entry and an array of entries for which the former entry is the context. Does that make sense?

bwindels replied:
Ah, this won't work because we only put the context entry in `_contextEntriesNotInTimeline` if it wasn't in the remoteEntries already, but in case of an update, we need to know which entries are using the context entry regardless of whether it came from remoteEntries or storage/hs.

bwindels replied:
Ok, never mind, let's keep it as it is :+1:

bwindels commented on src/matrix/room/timeline/entries/EventEntry.js:
```diff
@@ -39,6 +42,27 @@ export class EventEntry extends BaseEventEntry {
+        this._contextEntry = entry;
+    }
+
+    setAsContextOf(entry) {
```
can we make this method private and call it from `setContextEntry` above on the `entry` passed in? that way it's a little less noisy in Timeline...

MidhunSureshR replied:
We'd still want it to be public though:
```js
    setContextEntry(entry) {
        this._contextEntry = entry;
        entry.setAsContextOf(this);
    }

    setAsContextOf(entry) {
        if (!this._contextForEntries) {
            this._contextForEntries = [];
        }
        this._contextForEntries.push(entry);
    }
```

bwindels replied:
why is that? is it called from anywhere else now but setContextEntry?

MidhunSureshR replied:
nvm, I'm an idiot.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,7 +251,128 @@ export class Timeline {
+     */
+    _updateEntriesNotInTimeline(entries) {
+        for (const entry of entries) {
+            const relatedEntry = this._contextEntriesNotInTimeline.get(entry.relatedEventId);
```
This doesn't seem to take into account that a context entry that came from storage (but not in the timeline) will still be updated by the syncwriter, and the entry that is passed in will actually be the relation target.

So let's say we have a message E. Then there is a reply F to message E. E wasn't in the timeline, so is put into _contextEntriesNotInTimeline. Now a redaction G is sent to redact message E. When G is received by our client through sync, E is known in storage. So the sync writer, delegating to the relation writer, will fetch and update E from storage, and thus an already redacted version of E will be passed to replaceEntries, while G will be passed to addEntries.

I'm not sure it makes sense to call the same method from both addEntries and replaceEntries for this purpose. I think we should do what you already do for events that are not in storage (e.g. fetched from the homeserver) in addEntries, but we should also look for entry.id in _contextEntriesNotInTimeline for the above case in replaceEntries.

Does that make sense?

bwindels reviewed (commented):
Thanks for the PR description, that is very helpful!
A few more comments after another read of the code.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -206,16 +215,22 @@ export class Timeline {
     replaceEntries(entries) {
         this._addLocalRelationsToNewRemoteEntries(entries);
+        this._updateEntriesNotInTimeline(entries);
+        this._loadContextEntriesWhereNeeded(entries);
```
Not sure why we would need to load context after an event has been updated in any way (redacted mostly, and later on edited)? What could update an entry that changes the context entry that it needs? Relations can't be edited AFAIK...

MidhunSureshR committed:
Replace entry in contextEntryNotInTimeline

MidhunSureshR committed:
Move line into if

MidhunSureshR committed:
Keep calls internal to class

MidhunSureshR committed:
Refactor + put redaction in NonPersistedEventEntry

MidhunSureshR committed:
Fix tests

MidhunSureshR committed:
Move check down

MidhunSureshR committed:
use addLocalRelation

MidhunSureshR committed:
Rename method and add comment

MidhunSureshR committed:
Make setAsContextOf private

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,7 +254,111 @@ export class Timeline {
+
+    /**
+     * Update entries based on newly received events.
+     * This is specific to events that are not in the timeline but had to be fetched from the homeserver.
```
this comment on its own seems confusing - can you mention that the events had to be fetched from the homeserver because they are context entries for other events in the timeline?

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,7 +254,111 @@ export class Timeline {
+            if (relatedEntry?.addRelation(entry)) {
+                // update other entries for which this entry is a context entry
+                relatedEntry.contextForEntries?.forEach(e => {
+                    this._remoteEntries.findAndUpdate(te => te.id === e.id, () => { return { reply: relatedEntry }; });
```
I thought you didn't need reply in the end here as a param? You could just pass the string "contextEntry", the name of the property that changed, which is what is usually used as a param.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,7 +254,111 @@ export class Timeline {
+         */
+        for (const entry of entries) {
+            const relatedEntry = this._contextEntriesNotInTimeline.get(entry.relatedEventId);
+            if (relatedEntry?.addRelation(entry)) {
```
Won't this also add local relations for context entries that came from storage? Those as mentioned in the comment above will already be updated by replaceEntries, and could potentially see their relations double counted like this?

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,7 +254,111 @@ export class Timeline {
+            const fetchedEntry = this._contextEntriesNotInTimeline.get(entry.id);
+            if (fetchedEntry) {
+                fetchedEntry.contextForEntries.forEach(e => e.setContextEntry(entry));
+                entry.updateFrom(fetchedEntry);
```
Why do we need this? `contextForEntries` has been set by the previous line, not sure what else needs to be copied, as the entry that just came from the server has the latest event?

Also, to be completely consistent, we could also emit an update here for all the entries that have this entry as context? It likely wouldn't change anything on screen but it would ensure that the new entry is used everywhere down in the TilesCollection and the view, ...

MidhunSureshR replied:
Previously, `updateFrom` was needed to copy over `contextFor` to the new entry. Since we now do that implicitly in `setContextEntry()`, we can remove it.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,7 +254,111 @@ export class Timeline {
+        }
+    }
+
+    _updateEntry(entry) {
```
Would `_emitUpdateForEntry` be a better name? When I see it used elsewhere, I keep thinking that it updates the entry somehow inline, rather than just emitting to inform an update already happened.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,7 +254,111 @@ export class Timeline {
+     * @param {EventEntry[]} entries 
+     */
+    async _loadContextEntriesWhereNeeded(entries) {
+        const entriesNeedingContext = entries.filter(e => !!e.contextEventId);
```
Since you already have a loop below, it would be so easy to `break` with an `if` rather than to create a new array?

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,7 +254,111 @@ export class Timeline {
+     * @param {string} eventId event-id of the entry
+     * @returns entry if found, undefined otherwise
+     */
+    _getTrackedEntry(eventId) {
```
Still wonder if there is a less generic name for this ... what about `_findLoadedEventById`?

bwindels commented on src/matrix/room/timeline/entries/BaseEventEntry.js:
```diff
@@ -53,10 +53,13 @@ export class BaseEventEntry extends BaseEntry {
         @return [string] returns the name of the field that has changed, if any
     */
-    addLocalRelation(entry) {
+    addRelation(entry) {
```
The comment above is great, thanks, but not sure about this name change. It gives the impression that this is the only way relations are tracked, while 99% of the time, relations are tracked by RelationWriter, and this is just for some "edge" cases like local echo and when fetching something from /context.

If we can't come up with a better name (I'm not able to come up with one), I'd keep `addLocalRelation` tbh.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -216,6 +227,13 @@ export class Timeline {
         for (const entry of entries) {
             try {
                 this._remoteEntries.getAndUpdate(entry, Timeline._entryUpdater);
+                const oldEntry = this._contextEntriesNotInTimeline.get(entry.id)
```
I wonder what will happen if you reply to a pending message (e.g. while offline), and then everything gets sent when  off-line. `id` returns the transaction id for `PendingEventEntry`s and the event_id for `EventEntry`s. Could it happen that we are left with both the PendingEventEntry and the EventEntry for the same message in the map, as they use a different id?

bwindels replied:
Actually, in that scenario, the context entry will fail to load before it is sent, right? As:
 - it's not in remoteEntries
 - it's not in storage
 - it's not in the homeserver

bwindels replied:
Can you test this scenario? We might need to also support getting the context entry from `this._localEntries` as well.

MidhunSureshR replied:
Yes, the context entry would fail to load and this would render a `ReplyPreviewError`. When we're back online, the reply preview would be rendered. We need to extend ~~getTrackedEntry~~ findLoadedEventById to also look in localEntries for the context.

MidhunSureshR replied:
> I wonder what will happen if you reply to a pending message (e.g. while offline), and then everything gets sent when off-line. `id` returns the transaction id for `PendingEventEntry`s and the event_id for `EventEntry`s. Could it happen that we are left with both the PendingEventEntry and the EventEntry for the same message in the map, as they use a different id?

As we discussed earlier, this should probably never happen since by definition a pending message would always be in timeline and therefore the map is never involved.

bwindels reviewed (commented):
Thank you for your changes! Had some more thoughts while going over it another time.

MidhunSureshR committed:
Don't pass relatedEntry in param

MidhunSureshR committed:
Pass prop change

MidhunSureshR committed:
Remove dead code

MidhunSureshR committed:
Rename updateEntry -> emitUpdateForEntry

MidhunSureshR committed:
stylistic change

MidhunSureshR committed:
getTrackedEntry -> findLoadedEventById

MidhunSureshR committed:
update comment

MidhunSureshR committed:
Guard entry from storage being processed by method

MidhunSureshR committed:
Check in all entries for context

MidhunSureshR committed:
Fix looking in allEntries

MidhunSureshR committed:
Implement offline support for context entries

MidhunSureshR committed:
Fix tests

MidhunSureshR committed:
Remove statement

MidhunSureshR committed:
Restore earlier name

MidhunSureshR committed:
update contextForEntries

bwindels commented on src/matrix/room/timeline/entries/EventEntry.js:
```diff
@@ -142,18 +128,13 @@ export class EventEntry extends BaseEventEntry {
-    get contextEventId() {
-        if (this.isReply) {
-            return this.relatedEventId;
+    get threadEventId() {
```
this does not seem used?

MidhunSureshR replied:
It's used in PR-3 (createReplyContent)

MidhunSureshR replied:
Sorry, moved to PR-3

bwindels commented on src/matrix/room/timeline/entries/EventEntry.js:
```diff
@@ -142,18 +128,13 @@ export class EventEntry extends BaseEventEntry {
-        if (this.isReply) {
-            return this.relatedEventId;
+    get threadEventId() {
+        if (this.isThread) {
```
Also, this isn't defined anywhere?

MidhunSureshR replied:
sorry about this, but it seems to be added in PR-3?

MidhunSureshR replied:
moved this code to PR-3

bwindels commented on src/matrix/room/timeline/entries/EventEntry.js:
```diff
@@ -45,20 +43,8 @@ export class EventEntry extends BaseEventEntry {
-
-    get contextForEntries() {
-        return this._contextForEntries;
+    setRelatedEntry(entry) {
```
Same, this doesn't seem to be used?

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -751,6 +751,8 @@ export function tests() {
             const entryB = new EventEntry({ event });
             await timeline.load(new User(alice), "join", new NullLogItem());
+            timeline._setupEntries([]);
+            timeline._localEntries.onSubscribeFirst();
```
this seems a bit brittle ... consider subscribing to timeline.entries instead.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -751,6 +751,8 @@ export function tests() {
             }, createEvent("m.room.message", "event_id_2", bob));
             const entryB = new EventEntry({ event });
             await timeline.load(new User(alice), "join", new NullLogItem());
+            timeline._setupEntries([]);
```
I'd try to avoid calling private methods from tests if possible, as they can make the tests more brittle. Why not use load?

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -811,7 +817,12 @@ export function tests() {
             };
             const entryB = new EventEntry({ event: withContent(content, createEvent("m.room.message", "event_id_2", bob)) });
             const entryC = new EventEntry({ event: withContent(content, createEvent("m.room.message", "event_id_3", bob)) });
+            entryA._eventEntry.eventIndex = 1;
```
why not just set it in the object literal above?

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -141,6 +141,7 @@ export class Timeline {
             // redactingEntry might be a PendingEventEntry or an EventEntry, so don't assume pendingEvent
             const relatedTxnId = pee.redactingEntry.pendingEvent?.relatedTxnId;
             this._findAndUpdateEntryById(relatedTxnId, pee.redactingEntry.relatedEventId, updateOrFalse);
+            pee.redactingEntry.contextForEntries?.forEach(e => this._emitUpdateForEntry(e));
```
the reason `e` is emitting an update here is because it's contextEntry might have changed. Should we add a `param` parameter to `_emitUpdateForEntry` as an argument to the update event so we can specify `"contextEntry"` here (and elsewhere where we call `_emitUpdateForEntry`)?

MidhunSureshR replied:
yes, we do that for updates everywhere else 👍

bwindels reviewed (commented):
some more comments, sorry :)

MidhunSureshR committed:
Remove unused method

MidhunSureshR committed:
Clean test code; try not to peek into internals

MidhunSureshR committed:
add param to emitUpdateForEntry

MidhunSureshR committed:
Remove unused getter

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -304,6 +423,19 @@ export class Timeline {
+            // if allEntries isn't loaded yet, fallback to looking only in remoteEntries
+            return this.getByEventId(eventId);
+        }
+        for (const entry of this._allEntries) {
```
Given that we won't allow replying to unsent events, should we revert this to be just looking in remoteEntries now?

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -236,7 +256,117 @@ export class Timeline {
+            if (relatedEntry?.isNonPersisted && relatedEntry?.addLocalRelation(entry)) {
+                // update other entries for which this entry is a context entry
+                relatedEntry.contextForEntries?.forEach(e => {
+                    this._remoteEntries.findAndUpdate(te => te.id === e.id, () => "contextEntry");
```
can we use `_emitUpdateForEntry` here?

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -600,6 +734,177 @@ export function tests() {
+                }
+            }, createEvent("m.room.message", "event_id_2", bob));
+            const entryB = new EventEntry({ event, eventIndex: 2 });
+            timeline._getEventFromStorage = () => entryA
```
Ideally we'd put the entry in the mock storage, rather than monkey-patch a private method. But not a blocker. See the `local reaction removal gets applied after remote echo is added to timeline with reaction not loaded` test how this can be easily done.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -600,6 +734,177 @@ export function tests() {
+                }
+            }, createEvent("m.room.message", "event_id_2", bob));
+            const entryB = new EventEntry({ event, eventIndex: 2 });
+            timeline._getEventFromHomeserver = () => entryA
```
Ideally we'd pass a mock for the hsApi (with just the context method) rather than monkey-patch a private method, but not a blocker.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -600,6 +734,177 @@ export function tests() {
+                fragmentIdComparer, pendingEvents: new ObservableArray(), clock: new MockClock()});
+            const entryA = new NonPersistedEventEntry({ event: withTextBody("foo", createEvent("m.room.message", "event_id_1", alice)), eventIndex: 1 });
+            let event = withContent({
+                body: "bar",
```
feel free to put the  "m.relates_to" in a `withReply` method with the other event mock functions, up to you, not blocking.

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -600,6 +734,177 @@ export function tests() {
+            await poll(() => entryB.contextEntry);
+            const redactingEntry = new EventEntry({ event: withRedacts(entryA.id, "foo", createEvent("m.room.redaction", "event_id_3", alice)), eventIndex: 3 });
+            timeline.addEntries([redactingEntry]);
+            const contextEntry = timeline._contextEntriesNotInTimeline.get(entryA.id);
```
could we do `entryB.contextEntry.isRedacted` here to avoid accessing a private member, or you want to test if it's in entries without context as well?

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -600,6 +734,177 @@ export function tests() {
+            const bin = [entryB, entryC];
+            timeline.entries.subscribe({
+                onUpdate: (index) => {
+                    const i = bin.findIndex(e => e.id === index);
```
I'm not sure why we're comparing the entry id to the index here? Also, wouldn't it just be simpler to log the indices of the updates here and later assert that to `[1, 2]` or whatever the right value is?

MidhunSureshR replied:
Sorry, the logic for this test is borked. I've improved it somewhat. 
The problem with asserting indices is that update events may be emitted twice so the indices might be something like [0, 1, 0, 1].

bwindels commented on src/matrix/room/timeline/Timeline.js:
```diff
@@ -600,6 +734,177 @@ export function tests() {
+            assert.strictEqual(timeline._contextEntriesNotInTimeline.has(entryA.id), false);
+            const movedEntry = timeline.remoteEntries[0];
+            assert.deepEqual(movedEntry, entryA);
+            assert.deepEqual(movedEntry.contextForEntries.pop(), entryB);
```
Fine either way, but curious why you pop rather than `[0]`?

bwindels reviewed (approved):
Here's my final review, looking great! Just a few more small things, some of them up to you and not blocking. Thanks again for persisting at this!

MidhunSureshR committed:
check only in remoteEntries

MidhunSureshR committed:
Use emitUpdateForEntry

MidhunSureshR committed:
Override methods in NonPersistedEventEntry

This will prevent redactions to entries fetched from hs showing "message
is being redacted" and will instead show "message is redacted"

MidhunSureshR committed:
Use mock storage

MidhunSureshR committed:
Remove dead code from test

MidhunSureshR committed:
Use hsApi mock

MidhunSureshR committed:
Fix tests

MidhunSureshR committed:
Improve test logic

bwindels commented on src/matrix/room/timeline/entries/NonPersistedEventEntry.js:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    get isRedacted() {
+        return !!this._pendingRedactions;
```
```suggestion
        return super.isRedacting;
```

bwindels reviewed (commented):
small suggestion

MidhunSureshR committed:
use withReply

MidhunSureshR committed:
Return property from super

Co-authored-by: Bruno Windels <brunow@matrix.org>

bwindels reviewed (approved):
Thanks for the last changes, this is good to go!
